client/lib/cgroupslib/mount.go

```diff
@@ -27,7 +29,30 @@ func detect() Mode {
 		_ = f.Close()
 	}()
 
-	return scan(f)
+	mode := scan(f)
+	if mode == CG2 && !functionalCgroups2() {
+		return OFF
+	}
+	return mode
+}
+
+func functionalCgroups2() bool {
+	const controllersFile = "cgroup.controllers"
+	requiredCgroup2Controllers := []string{"cpuset", "cpu", "io", "memory", "pids"}
+
+	controllersRootPath := filepath.Join(root, controllersFile)
+	content, err := os.ReadFile(controllersRootPath)
+	if err != nil {
+		return false
+	}
+	rootSubtreeControllers := strings.Split(strings.TrimSpace(string(content)), " ")
+
+	for _, controller := range requiredCgroup2Controllers {
+		if !slices.Contains(rootSubtreeControllers, controller) {
+			return false
+		}
+	}
```

Comment:
This definitely works but to make this a bit cleaner you could do something like,

```go
rootSubtreeControllers := set.From[string](strings.Fields(string(content)))
return rootSubtreeControllers.ContainsSlice(requiredCgroup2Controllers)
```

(using https://pkg.go.dev/github.com/hashicorp/go-set/v2@v2.1.0#Set.ContainsSlice)